src/backend/vaapi/encoder.rs

```diff
@@ -505,7 +518,74 @@ pub(crate) mod tests {
         }
     }
 
-    pub fn upload_test_frame<M: SurfaceMemoryDescriptor>(
+    pub fn fill_test_frame_p010(
+        width: usize,
+        height: usize,
+        pitches: [u32; 3],
+        offsets: [u32; 3],
+        t: f32,
+        raw: &mut [u8],
+    ) {
+        let (sin, cos) = f32::sin_cos(t);
+        let (sin2, cos2) = (sin.powi(2), cos.powi(2));
+
+        // Pick the dot position
+        let dot_col = height as f32 * (1.1 + 2.0 * sin * cos) / 2.2;
+        let dot_row = width as f32 * (1.1 + sin) / 2.2;
+        let dot_size2 = (width.min(height) as f32 * 0.05).powi(2);
+
+        let mut dst = &mut raw[offsets[0] as usize..];
+
+        // Luma
+        for row in 0..height {
+            #[allow(clippy::needless_range_loop)]
+            for col in 0..width {
+                let dist = (dot_col - col as f32).powi(2) + (dot_row - row as f32).powi(2);
+
+                let y = if dist < dot_size2 {
+                    0
+                } else {
+                    0x3ff * (row + col) / (width + height)
+                };
+
+                dst[2 * col + 0] = ((y << 6) & 0xa0) as u8;
+                dst[2 * col + 1] = (y >> 2) as u8;
+            }
+
+            dst = &mut dst[pitches[0] as usize..];
+        }
+
+        // Advance to the offset of the chroma plane
+        let mut dst = &mut raw[offsets[1] as usize..];
+
+        // Chroma
+        for row in 0..height / 2 {
+            let row = row * 2;
+
+            for col in 0..width / 2 {
+                let col = col * 2;
+                let dist = (dot_col - col as f32).powi(2) + (dot_row - row as f32).powi(2);
+
+                let c = if dist < dot_size2 {
+                    (512, 512)
+                } else {
+                    (
+                        (((row * 0x3ff) / width) as f32 * sin2) as u32,
+                        (((col * 0x3ff) / height) as f32 * cos2) as u32,
+                    )
+                };
+
+                dst[2 * col + 0] = ((c.0 << 6) & 0xa0) as u8;
+                dst[2 * col + 1] = (c.0 >> 2) as u8;
+                dst[2 * col + 2] = ((c.1 << 6) & 0xa0) as u8;
+                dst[2 * col + 3] = (c.1 >> 2) as u8;
+            }
```

Comment:
I think we can avoid duplicating the code if we make this a function that takes a closure to write the color values. Just an idea for a future CL.